### ISSUE TITLE
Different behavior with repeated find_glycan() calls 

### ISSUE THREAD

**Issue by meteosR** [ibmm-unibe-ch/glycosylator#4]
Hello, `find_glycan()` seems to give some issues:

1. The behavior changes with multiple iterations
2. The first time it runs, it only finds 7 out of the 8 monosaccharides 
3. The second iteration, it finds all 8 monosaccharides, but removal of those glycans doesn't work (maybe similar to last issue https://github.com/ibmm-unibe-ch/glycosylator/issues/3#issue-2446146009) 

I am running the latest dev glycosylator version: https://github.com/ibmm-unibe-ch/glycosylator.git@dev

```
import glycosylator as gl
import glycosylator.resources as resources
import buildamol as bm

ref_residues_glycans = resources.reference_glycan_residue_ids()
path_in = "8F0P.cif"

print("gl", gl.__version__)
print("bm", bm.__version__)

for _ in range(2):
    try:
        print("\n\n")
        # Load structure
        scaffold = gl.protein(path_in)

        all_glycan_residues = set([r for r in scaffold.get_residues() if r.resname in ref_residues_glycans])
        print("all_glycan_residues", all_glycan_residues)

        # Find glycans
        glycans = scaffold.find_glycans()

        # Find number of glycans parsed
        num_residues = sum([len(list(v.get_residues())) for v in glycans.values()])

        print("num_residues", num_residues)
        print("all_glycan_residues", len(all_glycan_residues))
        assert num_residues == len(all_glycan_residues), "Monosaccharide mismatch"

        print("Removing glycans...")
        for g in glycans.values():
            scaffold.remove_glycan(g)

        all_glycan_residues = set([r for r in scaffold.get_residues() if r.resname in ref_residues_glycans])
        print("\nGlycan residues after removal:", all_glycan_residues)
        assert len(all_glycan_residues) == 0, "Not all monosaccharides were removed"

    except AssertionError as e:
        print("Error: ", e)
```

Here is my output 
<img width="1040" alt="Screenshot 2024-08-09 162327" src="https://github.com/user-attachments/assets/baaca0a4-9a56-4568-8acc-ae85d4e30a0b">

[8F0P.txt](https://github.com/user-attachments/files/16566378/8F0P.txt)

**Comment by NoahHenrikKleinschmidt** [comment]
You again 😄, I'm happy you are trying out what Glycosylator can do and find some good issues, thanks! 

> ### TLDR
> Yes, there is a stochastic process involved in the inference, so it is possible that certain glycans may fail to be fully recovered. I added an additional round of iterations to the process, however, to make it more likely that the inference is successful. However, there is no 100% guarantee.
>
> The update is now available via the `dev` branch. When I tried with your example, I could recover the full connectivity. See also the long explanation for more details :-)

<details>
<summary><h3>The long explanation</h3></summary>

So the reason why the behavior appears to change is that the glycan finding involves a random step in which the connectivity of glycans are inferred through repeatedly sampling through a `set` of so-far unassigned glycan residues. It is possible that the sampling is unable to assign the residues in the right order due to that random representation which is more memory efficient than using an ordered representation. 

I ran the same loop of inference as you for 50 times and counted how many residues were found in the "long" glycan. I found that with the original set implementation 1 in 5 runs could not assign the full connectivity because the `BMA@3` was sampled later than the subsequent residues that would have dependent on it for full reconstruction (see figure below). 

```python
n_residues = []
for _ in range(50):
   P = gl.Protein.form_cif("~/Downloads/8F0P.cif")
   P.find_glycans()
   n_residues.append(P.glycans[-1].count_residues())

plt.hist(n_residues)
```

<img width="437" alt="image" src="https://github.com/user-attachments/assets/ce111df3-970f-454a-ba44-beb380336e0d">

I then changed the implementation to use a `list` (ordered)  instead of a `set` (unordered/random), which improved the results a little. However, there were still runs where the full connectivity was not obtained. 

Instead, the issue primarily arises ("arose",  I should say now) from the fact that Glycosylator only iterated a single time over the residues for which the connectivity was inferred but dependent on a residue that was not yet added to a Glycan object. Therefore, even though the connectivity was inferred (it was always inferred correctly, actually), it could not be applied to Glycan objects because of the sampling order. By now iterating over these "preemptively assigned" residues multiple times the chance is much higher that the full connectivity is recovered. 

When I tried with your input, all 50 times the glycan was fully recovered now.

```python
import matplotlib.pyplot as plt
fig, axs = plt.subplots(2, 2, figsize=(8, 8))
for anchor, ax in zip(P.get_glycans(), axs.flat):
   g = P.get_glycan(anchor)
   g.snfg(ax=ax)
   ax.set_title(g.id + " @ " + f"{anchor.parent.resname}-{anchor.parent.serial_number}")
   ax.axis("on")
   ax.set_xticks([])
   ax.set_yticks([])
```

![image](https://github.com/user-attachments/assets/121fdaf8-8cf9-4175-8ec2-5bb635664223)

</details>

Regarding the residues that could not be removed, that is linked to them being not properly assigned into Glycan objects, so this should be fixed as well. When I tried, I could remove them all seamlessly. 

By the way, if you are only interested in removing the glycans you don't actually have to infer them, you can simply use `remove_residues` to get rid of them like so:

```python
P = gl.Protein.from_cif("~/Downloads/8F0P.cif")

# get the glycan residues without inferring the glycan structures etc.
glycan_residues = P.get_residues(*gl.reference_glycan_residue_ids())

# now remove them
P.remove_residues(*glycan_residues)
```

Like this you avoid doing the inference computations which are not really relevant for removing the glycans 😄

**Comment by meteosR** [comment]
Thank you very much, the new code indeed finds the missing monosaccharides!  

However, I am still having issues removing the glycans. Running the same code from above, I failed my second assertion ("Not all monosaccharides were removed"). 

If I try your suggested method (not infering the glycan trees) I get the following error:

<img width="862" alt="8_12_24" src="https://github.com/user-attachments/assets/a0e52bad-ef66-4ab2-8477-087dab8db0ce">

**Comment by NoahHenrikKleinschmidt** [comment]
Hmmm, okey, I found what was causing the problem. And works fine for me now.  

The issue was that if casting the glycans to the same chain as the anchor atom, the glycans were not detached from their original chain (assuming they would be in the same chain as the scaffold-anchor atom anyway), but this was not the case for your input file where the "long" glycan was in a separate chain B, while the rest of the protein (and the small glycans too) were in chain A. Usually, the glycan inference should handle the detaching, but I missed that part when adding the extra inference loop to make the whole thing more robust. Now it's back in the code. 

<details>
<summary>Details: Chain Casting</summary>

When using `Scaffold.find_glycans` one can provide a `chain` argument which can be either `same` (default), `new` or `each`. When keeping it as `same` each glycan is considered part of the same chain as the anchor atom. If set to `new` then all glycans are considered to form a new chain together - this would be a discontinuous chain, of course. If set to `each` each glycan creates it's own chain in the scaffold. 

I usually use `new` instead of `same` when inferring glycans, since it makes things cleaner :-) 

</details>

About the List attribute error: just unpack it like so `P.remove_residues(*residues)` and it will work (I updated the example above too). My bad, originally BuildAMol is designed to work "for convenience" with unpacked inputs such as `P.remove_residues(1, 2, 3)` rather than collections such as `P.remove_residues([1, 2, 3])`. The newer versions have added more and more support for this as well, however, making it so that you can use either syntax and it's going to work. It's true that `remove_residues` did not support this in `v.1.2.6` yet, but you could clone and install from the `dev` branch of BuildAMol to make the behaviour work, if you like 😄

**Comment by meteosR** [comment]
That's great, thanks for all the fixes and detailed explanations, really appreciate it !
